### ISSUE TITLE
Bug in geiger_reigime.cc / mock_tracker_s2c

### ISSUE THREAD

**Issue by edbird** [SuperNEMO-DBD/Falaise#80]
I think I've found a bug. Here is how I came to this conclusion:

I produced a histogram of the cathode_time produced by Falaise after processing the output with my own modules. I then uncommented a line of code in my modules.defs file to set the _plasma_propagation_time_ variable. I left it at the default value. I re-ran the entire sequence from data generation to processing and the output histogram contained results which were scaled by a factor of 1000.0. It's possible the bug is in my own code, however here is why I don't believe this to be the case:

I modified the geiger_regime file:

```
--- a/source/falaise/snemo/processing/geiger_regime.cc
+++ b/source/falaise/snemo/processing/geiger_regime.cc
@@ -82,10 +82,13 @@ void geiger_regime::initialize(const datatools::properties& config_) {
 
   if (config_.has_key("plasma_longitudinal_speed")) {
     _plasma_longitudinal_speed_ = config_.fetch_real("plasma_longitudinal_speed");
+    std::cout << "config has key: pls=" << _plasma_longitudinal_speed_ << std::endl;
     if (config_.has_explicit_unit("plasma_longitudinal_speed")) {
       _plasma_longitudinal_speed_ *= drift_speed_unit;
+      std::cout << "config has unit: pls=" << _plasma_longitudinal_speed_ << std::endl;
     }
   }
+  std::cout << "regardless: pls=" << _plasma_longitudinal_speed_ << std::endl;
 
   if (config_.has_key("sigma_plasma_longitudinal_speed")) {
     _sigma_plasma_longitudinal_speed_ = config_.fetch_real("sigma_plasma_longitudinal_speed");
```

If I ENABLE the line of code in modules.defs to set the PLS (plas. long. speed) variable, this output is produced:

has key 0.05
has unit 0.005
regardless pls=0.0005
regardless pls=0.05

If I DISABLE THAT LINE: This output is produced.

regardless pls=0.05
regardless pls=0.05

I cannot figure out why it appears to be the case that the module (geiger_regime) is loaded TWICE. Unless it is used somewhere other than mock_tracker_s2c ?

Regardless of this; in the ENABLED example: The PLS is set to 0.0005 as expected, but it is then somehow set back to 0.05?

-> If I am right in my assumption that this module (geiger) is loaded by 2 different parts of flreconstruct, then it must be the case that the second bit of code which does the loading does not point geiger_regime at the correct file containing the config variable plasma_longitudinal_speed : real as speed = 5.0 cm/us

But this is just my guess as what is going on.

Sorry this is not the most logically worded explanation - we can perhaps discuss it further via slack?

**Comment by drbenmorgan** [comment]
> I cannot figure out why it appears to be the case that the module (geiger_regime) is loaded TWICE. Unless it is used somewhere other than mock_tracker_s2c ?

Yes, it's also used in the TrackFit module: https://github.com/SuperNEMO-DBD/Falaise/blob/develop/modules/TrackFit/source/falaise/snemo/reconstruction/trackfit_driver.h

It doesn't configure it though, hence the difference in the plasma speed. I haven't tracked through `trackfit_driver` to see how critical this is though.

**Comment by drbenmorgan** [comment]
I've assigned @xgarrido @fmauger and @fmauger initially as the had the main authorship of the Trackfit module. 

I guess the question is "where does TrackFit use `geiger_regime` and should its parameters be consistent between each place it's used?". Feels like it should be and needs input from the Calibration group?

**Comment by edbird** [comment]
If it's as simple as loading that parameter from the config, exactly as it is done in the mock_tracker_s2c_module, then I could fork the current version of falaise (not my own version but the development branch in this repo) and add in the code to do this and submit a PR. Would it be helpful if I was to do this?

**Comment by edbird** [comment]
Actually that's not helpful at all since this module doesn't seem to use that variable.

**Comment by drbenmorgan** [comment]
I think we need clarification on use of `geiger_regime` first!

**Comment by edbird** [comment]
I think what might be happening is that the units in the modules.defs file does not match up with the units expected by Falaise.

cm/us seems to have a value of 0.001.

drift_speed_unit is CLHEP::cm / CLHEP::microsecond

If no value is provided, the default is used. This is: `_plasma_longitudinal_speed_ = 5.0 * CLHEP::cm / CLHEP::microsecond;`

If a value is provided in the config file, eg: "5.0 cm/us" then I can't be sure what happens because I don't know how this is read.

I do not understand the line below: `if (config_.has_explicit_unit("plasma_longitudinal_speed"))` - is it not possible for any other unit is specified other than "cm/us" ?

Sorry if this is just spam by the way - I'm trying to give as much info to point to where the issue might be.

**Comment by xgarrido** [comment]
The `geiger_regime` is used in `Trackfit` only when recalibration of geiger hit is needed *i.e.* when a delayed alpha track is fitted. With delayed hits, the track fitting do a recalibration of geiger cells and thus needs one function from `geiger_regime` : `calibrate_drift_radius_from_drift_time` (https://github.com/SuperNEMO-DBD/Falaise/blob/develop/modules/TrackFit/source/falaise/snemo/reconstruction/trackfit_driver.cc#L52)

This function does not use the plasma longitudinal speed at all whose value is only needed for z calibration.

**Comment by xgarrido** [comment]
You should set something like that in your `module.def` file

```
  #@description plasma longitudinal speed
  plasma_longitudinal_speed : real as velocity = 5.0 cm/us
```

The `has_explicit_unit` method force the user to set velocity value with explicit unit otherwise default are given.

**Comment by edbird** [comment]
I will add two output histograms to demonstrate.

Firstly, here is that line disabled:

![c_cathode_time_default](https://user-images.githubusercontent.com/2939755/35156982-74895d20-fd2a-11e7-98e6-ce2440a16399.png)

And now enabled...
![c_cathode_time_en_5 0_cm_us](https://user-images.githubusercontent.com/2939755/35156992-7bfd7e60-fd2a-11e7-882c-953a794bb185.png)

Here's a more zoomed out version of the above.


![screenshot_2018-01-19_15-08-14](https://user-images.githubusercontent.com/2939755/35157051-b5dce652-fd2a-11e7-9954-45a9b060df7e.png)

**Comment by edbird** [comment]
Just ignore the "fits" they are wrong anyway.

**Comment by drbenmorgan** [comment]
@edbird , I _think_ (it's Friday afternoon...) there is a bug here. If we compare the logic in how the keys and values are extracted between `sigma_cathode_time` and `plasma_propagation_speed`

https://github.com/SuperNEMO-DBD/Falaise/blob/36806ea1261b4bb714edf300ac3a196b0d739097/source/falaise/snemo/processing/geiger_regime.cc#L68-L88

they are different in the check on the presence of the unit. In the case of `plasma_propagation_speed`, it looks like the value gets multiplied by the default unit _even_ if the value had a unit applied. 

@xgarrido, @fmauger could you confirm if that's the case?

**Comment by xgarrido** [comment]
You are right @drbenmorgan . I just proposed a quick fix in PR #81 but I also think your pseudo code will be better. I do not like the idea to leave the user with a default unit. If the user decides to change a parameter then he has to take care of the unit. Otherwise your `fetch_real_with_explicit_dimension` must throw an exception. Since the module configuration are full of these kind of check, it is better to change them in another (big) PR.

**Comment by edbird** [comment]
Ah I should have been able to spot that - the issue is with the true/false inversion due to the presence of `!` ?

I would suggest it would be a good idea for the user to be able to configure both the value and the unit. So, if a value is supplied, a unit should be mandatory, and if the user does not provide one, a terminal error should be thrown.

I don't think it would be a good idea to "assume" a default unit in the case that one is not provided - this is the kind of thing that would cause me to make a mistake, as I am the sort of person who would assume the unit was "SI" m/s when it isn't.
